### PR TITLE
Tx ID as Asset ID

### LINKED ISSUE #697: Use transaction ID as asset ID
When an asset is created, it needs to be identifiable with a unique ID, which can identify the set of transactions of this asset.

This ID can't be a hash of the asset, as the asset may have no distinguishing properties to give it a unique hash. The client may be trusted to generate a random asset ID, but this is also a potential attack vector and makes it more complicated to validate a block.

Using the ID of the transaction that created the asset has some nice properties:
- Client can still generate valid transactions
- No extra validation step is required
- No potential to create a duplicate (without privkey compromise), because the transaction ID depends on ~~a signature of the person creating  the asset~~ the public keys of the people creating the asset, verified by the signatures.

Thoughts?

### LINKED ISSUE #708: Why use "id" and "data_id" for the same thing?
_From @sbellem on October 18, 2016 16:29_

``` python
class Asset(object):
    def __init__(self, data=None, data_id=None, divisible=False,
                 updatable=False, refillable=False):
        self.data = data
        self.data_id = data_id if data_id is not None else self.to_hash()
        self.divisible = divisible
        self.updatable = updatable
        self.refillable = refillable

    def to_dict(self):
        return {
            'id': self.data_id,
            'divisible': self.divisible,
            'updatable': self.updatable,
            'refillable': self.refillable,
            'data': self.data,
        }
```

doing so prevents one from passing directly the results of `Asset.to_dict()` to the Python built-in class initialization

_Copied from original issue: bigchaindb/bigchaindb-common#58_

### LINKED ISSUE #715: Revist `to_hash()` in Assets and Metadata
_From @sohkai on October 13, 2016 8:53_

Assets and metadata don't have hashes associated with them (yet?), so the naming of `to_hash()` to get a uuid is somewhat misleading.

_Copied from original issue: bigchaindb/bigchaindb-common#43_

### LINKED ISSUE #717: Rename `data_id` in the Metadata and Asset models
_From @r-marques on October 10, 2016 15:25_

I find the name `data_id` confusing in both the `Metadata` and `Asset` model.
Since they are already inside a namespace I propose to rename `data_id` to `id`.
This way we have:
- `Transaction.id`
- `Transaction.metadata.id`
- `Transaction.asset.id`

_Copied from original issue: bigchaindb/bigchaindb-common#39_

### PR DESCRIPTION
**Pull request by libscott**
Use the ID of the transaction that CREATEd the asset to refer to the asset.

**Note**: with these changes, you'll now need to provide either a unique asset or metadata payload (along with the list of signers) to create a unique TX ID. If you're going to be creating the same asset without a unique metadata payload, you should probably be thinking about using a divisible asset (and a refillable one, later).

- [x] Make code changes to use TX ID to refer to Asset ID
- [x] Fix tests
- [x] Docs changes (@ttmc I wasn't able to find too much that had to be changed with this change in the server docs, maybe you know of more places?)

Fixes #697, fixes #717. fixes #715, fixes #708.

### CODE REVIEW

File: bigchaindb/common/schema/transaction.yaml [hunk 1]
```diff
@@ -105,13 +105,14 @@ definitions:
     description: |
         Description of the asset being transacted. In the case of a ``TRANSFER``
         transaction, this field contains only the ID of asset. In the case
-        of a ``CREATE`` transaction, this field may contain properties:
+        of a ``CREATE`` transaction, this field contains only a user payload
+        associated with the asset.
     additionalProperties: false
-    required:
-    - id
```
Comment (sohkai) [inline bigchaindb/common/schema/transaction.yaml]: Looked around to see if we could change the asset model based on `CREATE` and `TRANSFER` transaction but [it'd look pretty ugly](https://github.com/json-schema/json-schema/issues/158#issuecomment-86124035) and might be slower.

Thoughts @libscott? Right now the asset schema is a bit of a lie ;).

Comment (libscott) [inline bigchaindb/common/schema/transaction.yaml]: I'm actually in favour of using the oneOf approach. That way it can match both `{operation:CREATE, asset:{...}}` and `{operation:TRANSFER, asset_id:...}}`. But the more complicated these constraints become, the worse the schema will perform...

Comment (sohkai) [inline bigchaindb/common/schema/transaction.yaml]: Turns out there's a few other places where we're generalizing properties to fit different use cases (e.g. `CREATE`'s inputs).

This is a fairly different problem at hand, so we'll leave it for another issue: https://github.com/bigchaindb/bigchaindb/issues/946.

File: tests/common/test_transaction.py [hunk 2]
```diff
@@ -481,6 +475,15 @@ def test_asset_link_deserialization_with_empty_payload():
     assert asset_link == expected
 
 
+def test_asset_link_from_inputs(utx, transfer_utx):
+    from bigchaindb.common.transaction import AssetLink
+
+    expected = AssetLink(utx.id)
+    asset_link = AssetLink.from_inputs(utx, transfer_utx)
```
Comment (libscott) [inline tests/common/test_transaction.py]: This appears to be the only time that `from_inputs` is used with multiple arguments. Is it neccesary?

File: bigchaindb/common/transaction.py [hunk 3]
```diff
@@ -508,20 +503,20 @@ def validate_asset(self, amount=None):
                                   ' greater than one')
 
 
-class AssetLink(Asset):
+class AssetLink(object):
```
Comment (libscott) [inline bigchaindb/common/transaction.py]: Is this guy still needed? What about just having a dict rather than a whole object to store an ID?

**Comment by libscott** [comment]
Two possibilities:

1: polymorphic 'asset' key
```
{operation: CREATE, asset: {data: blah}}
{operation: TRANSFER, asset: {id: abc}}
```

2: separate keys
```
{operation: CREATE, asset: {blah}}
{operation: TRANSFER, asset_id: abc}
```

Interested to hear reasoning one way or the other.

But, from a logical point of view they are the same and it's not clear that we need the `Asset` or `AssetLink` classes anymore. Or at least, we won't once we remove the keywords from the `Asset`.

**Comment by diminator** [comment]
are we also considering squatting of the `tx_id`? ie: what happens if someone makes 2 separate `CREATE` txs with the same content? Will they yield the same `tx_id` or do we need to force the user to provide a `UUID`?

File: bigchaindb/common/transaction.py (hunk 3 above)
Comment (sohkai) [inline bigchaindb/common/transaction.py]: Agreed, but let's move this to another PR after #962 is merged so that we can clean up both.

File: tests/common/test_transaction.py (hunk 2 above)
Comment (sohkai) [inline tests/common/test_transaction.py]: Not really... but then again, none of the common stuff is actually necessary in the server-side code (ala @sbellem's realization that changing the common's API is actually more driver-related work rather than server-related).

**Comment by sohkai** [comment]
@diminator: See note on https://github.com/bigchaindb/bigchaindb/issues/697#issuecomment-267335527.

We basically decided the unique factor must either come in the metadata or asset, otherwise you should just be making divisible assets.

**Comment by sohkai** [comment]
@libscott For the `asset` key, I think it's more elegant if we make it polymorphic. That way the root structure of a transaction stays the same between the two types.

File: bigchaindb/common/transaction.py (hunk 3 above)
Comment (libscott) [inline bigchaindb/common/transaction.py]: I'm thinking it could be a good idea to bust out some `namedtuple`s :smiling_imp:

File: tests/common/test_transaction.py (hunk 2 above)
Comment (libscott) [inline tests/common/test_transaction.py]: The API here looks a bit odd to me. It's not clear that you're supposed to pass in all your transactions so that it can validate that they have the same asset id.

Comment (sohkai) [inline tests/common/test_transaction.py]: Fair enough, I guess this is actually doing more validation than building anything 😉 . I'll take a look into the driver to see how it uses `Transaction.transfer()` and try to remove this.

**Comment by libscott** [comment]
LGTM

File: bigchaindb/common/schema/transaction.yaml [hunk 4]
```diff
@@ -105,13 +105,14 @@ definitions:
     description: |
         Description of the asset being transacted. In the case of a ``TRANSFER``
         transaction, this field contains only the ID of asset. In the case
-        of a ``CREATE`` transaction, this field may contain properties:
+        of a ``CREATE`` transaction, this field contains only the user-defined
+        payload and asset-related keywords.
     additionalProperties: false
-    required:
-    - id
```
Comment (r-marques) [inline bigchaindb/common/schema/transaction.yaml]: The id is still required for `TRANSFER` transactions, right?

Comment (sohkai) [inline bigchaindb/common/schema/transaction.yaml]: It is, but we don't currently have a good way to split the requirements based on the operation (see #946).

File: bigchaindb/common/transaction.py [hunk 5]
```diff
@@ -575,10 +575,17 @@ def __init__(self, operation, asset, fulfillments=None, conditions=None,
...
-            raise TypeError('`asset` must be an Asset instance')
+        # Assets for 'CREATE' and 'GENESIS' operations must be None or of Asset
+        # type and Assets for 'TRANSFER' operations must be of AssetLink type.
+        if (operation in [Transaction.CREATE, Transaction.GENESIS] and
+                asset is not None and
+                not isinstance(asset, Asset)):
+            raise TypeError(("`asset` must be an Asset instance for "
+                             "'{}' Transactions".format(operation)))
+        elif (operation == Transaction.TRANSFER and
+                not (asset and isinstance(asset, AssetLink))):
```
Comment (r-marques) [inline bigchaindb/common/transaction.py]: This can be simplified to `asset not isinstance(asset, AssetLink)`. It already returns `True` if asset is `None`

File: bigchaindb/common/transaction.py [hunk 6]
```diff
@@ -575,10 +575,17 @@ def __init__(self, operation, asset, fulfillments=None, conditions=None,
...
-                not asset and operation != Transaction.CREATE):
-            raise TypeError('`asset` must be an Asset instance')
+        # Assets for 'CREATE' and 'GENESIS' operations must be None or of Asset
+        # type and Assets for 'TRANSFER' operations must be of AssetLink type.
+        if (operation in [Transaction.CREATE, Transaction.GENESIS] and
+                asset is not None and
+                not isinstance(asset, Asset)):
+            raise TypeError(("`asset` must be an Asset instance for "
+                             "'{}' Transactions".format(operation)))
+        elif (operation == Transaction.TRANSFER and
```
Comment (r-marques) [inline bigchaindb/common/transaction.py]: This is a big breaking change. `AssetLink` was something that only add meaning in the server code.
Clients would always pass an asset regardless and we would serialized it according to the operation.

Its probably worth it to open an issue in the bigchaindb-driver

Comment (sohkai) [inline bigchaindb/common/transaction.py]: @sbellem and I had a quick conversation about this change, since it'll affect the driver most. There's [this issue](https://github.com/bigchaindb/bigchaindb-driver/issues/202) for fixing the driver once this is merged.

File: bigchaindb/common/transaction.py (hunk 5 above)
Comment (sohkai) [inline bigchaindb/common/transaction.py]: Did you mean:

```py
elif (operation == Transaction.TRANSFER and not isinstance(asset, AssetLink)):
    ...
```

?

The `not asset` part is still important to keep because we want to check that the AssetLink actually contains a link (with its bool operator), although this is not entirely obvious.

Comment (r-marques) [inline bigchaindb/common/transaction.py]: Ok, I guess that is on way of doing it.
